### ISSUE TITLE
Subscription Component: "Can't perform a React state update on an unmounted component." when onSubscriptionData triggers a parent's setState and unmounts the Subscription

### ISSUE THREAD

**Issue by yzxj** [apollographql/react-apollo#2681]
**Intended outcome:**
I have a modal that contains two components, an input field and the Subscription component, which is listening to a single entry in the database for input from another device. Upon receiving an update, the component modal will unmount both components to display a spinner, blocking further inputs. This is done by passing down a binded function containing the modal's `setState` to the Subscription's `onSubscriptionData`. The rough structure follows this pseudocode:
```js
// Rough structure, not functional
class MyModal extends React.Component {
	state = { loading: false }

	setToLoading = () => this.setState({ loading: true })
	
	render() {
		const { loading } = this.state;
		if (loading) {
			return (<Modal><Spinner /></Modal>);
		}
		return (
			<Modal>
				<InputField />
				<Subscription onSubscriptionData={this.setToLoading} />
			</Modal>
		);
	}
}
```
The intended outcome is that there's no error output in the browser console indicating a "memory leak".

**Actual outcome:**
The UI renders properly (i.e. `setToLoading` runs, and the modal switches to the spinner. However, the browser console will display the following error.
![image](https://user-images.githubusercontent.com/7572120/50323389-6bf1bc00-0514-11e9-9595-ef238c8f3f4a.png)

**How to reproduce the issue:**
I'm not sure this small issue warrants a full sandbox reproduction. In any case, I believe I may have found the cause. After some debugging, I reached this section of code in react-apollo:
```js
// Inside src/Subscriptions.tsx
  private updateCurrentData = (result: SubscriptionResult<TData>) => {
    const {
      client,
      props: { onSubscriptionData },
    } = this;
    if (onSubscriptionData) onSubscriptionData({ client, subscriptionData: result });
    this.setState({
      data: result.data,
      loading: false,
      error: undefined,
    });
  };
```
In this code, `onSubscriptionData` is run before `this.setState`. This means the parent component's `setState` gets run before the Subscription's. When I switched the order, my error went away.

**Version**

- apollo-client@2.4.7
- react-apollo@2.3.3

**Comment by slikts** [comment]
I believe I've run into the same issue, except using `Query` and `subscribeToMore()`.

**Comment by kardoqadir** [comment]
The same here...

**Comment by avasuro** [comment]
Same story

**Comment by clairefritz** [comment]
Same here

**Comment by scottie-schneider** [comment]
This is really ruining my evening as well lol

**Comment by MikeAlexMartinez** [comment]
Me too

**Comment by sorokinvj** [comment]
any update on this? 

Could I unsubscribe from <Subscription> in componentWillUnmount as React warning suggests?

**Comment by avasuro** [comment]
@sorokinvj 

There is no possibility to unsubscribe in componentWillUnmount without changing apollo source code

**Comment by sorokinvj** [comment]
> @sorokinvj
> 
> There is no possibility to unsubscribe in componentWillUnmount without changing apollo source code

ok, so I guess I should take another approach with subscribeToMore

**Comment by jasonpaulos** [comment]
Hi everyone! A fix for this issue has been merged with the codebase. It will be included in the next release of react-apollo.

**Comment by darrylyoung** [comment]
Has this updated version already been released, @jasonpaulos? Thank you.

**Comment by jasonpaulos** [comment]
@darrylyoung yes, I believe this fix was included in react-apollo 2.5.7.
